code/modules/mob/living/simple_animal/hostile/megafauna/chaos_marine.dm

```diff
@@ -0,0 +1,510 @@
+	playsound(src,'sound/effects/meteorimpact.ogg', 200, 1)
+
+/mob/living/simple_animal/hostile/megafauna/chaos_marine/proc/blood_dash_2(var/move_dir, var/times_ran)
+	if(times_ran >= dash_num)
+		charging = FALSE
+		return
+	var/turf/T = get_step(get_turf(src), move_dir)
+	if(ismineralturf(T))
+		var/turf/closed/mineral/M = T
+		M.gets_drilled()
+	if(T.density)
+		charging = FALSE
+		return
+	for(var/obj/structure/window/W in T.contents)
+		charging = FALSE
+		return
+	for(var/obj/machinery/door/D in T.contents)
+		if(D.density)
+			charging = FALSE
+			return
+	forceMove(T)
+	playsound(src,"sound/effects/footstep/heavy[pick(1,2)].ogg", 200, 1)
+	for(var/mob/living/L in T.contents - src)
+		if(!faction_check_mob(L))
+			visible_message("<span class='boldwarning'>[src] runs through [L]!</span>")
+			to_chat(L, "<span class='userdanger'>[src] pierces you with a chain-sword!</span>")
+			explosion(L, -1, 0, 2, 0, 0, flame_range = 0)
+			shake_camera(L, 4, 3)
+			playsound(L,"sound/effects/wounds/pierce[pick(1,2,3)].ogg", 200, 1)
+	addtimer(CALLBACK(src, .proc/blood_dash_2, move_dir, (times_ran + 1)), (1.5 * dash_mod))
+
+/mob/living/simple_animal/hostile/megafauna/chaos_marine/proc/teleport_b(target)
+	if(charging || teleport_cooldown > world.time)
+		return
+	charging = TRUE
+	telegraph()
+	cmempower()
+	var/turf/T = get_step(target, pick(1,2,4,8))
+	var/turf/S = get_turf(src)
+	if(T.density)
+		T = get_turf(target)
+	layer = 3
+	SLEEP_CHECK_DEATH(4)
+	new /obj/effect/temp_visual/cult/blood(S)
+	SLEEP_CHECK_DEATH(1)
+	new /obj/effect/temp_visual/cult/blood/out(T)
+	SLEEP_CHECK_DEATH(2)
+	forceMove(T)
+	playsound(T,'sound/magic/exit_blood.ogg', 200, 1)
+	if(health <= maxHealth*0.5)
+		layer = initial(layer)
+		cmdepower()
+		teleport_cooldown = world.time + (initial(teleport_cooldown) * dash_mod)
+		charging = FALSE
+		blood_dash(target)
+	else
+		SLEEP_CHECK_DEATH(5 * dash_mod)
+		layer = initial(layer)
+		cmdepower()
+		teleport_cooldown = world.time + (initial(teleport_cooldown) * dash_mod)
+		charging = FALSE
+
+/mob/living/simple_animal/hostile/megafauna/chaos_marine/proc/blast()
+	if(ranged_cooldown <= world.time && !Adjacent(target) && !charging)
+		ranged_cooldown = world.time + ranged_cooldown_time
+		visible_message("<span class='danger'>[src] fires the bolter gun!</span>")
+		face_atom(target)
+		new /obj/effect/temp_visual/dir_setting/firing_effect(loc, dir)
+		Shoot(target)
+		changeNext_move(CLICK_CD_RANGE)
+
+/mob/living/simple_animal/hostile/megafauna/chaos_marine/proc/rapid_fire()
+	if(ranged_cooldown <= world.time && !Adjacent(target) && !charging)
+		ranged_cooldown = world.time + ranged_cooldown_time
+		charging = TRUE
+		visible_message("<span class='danger'>[src] rapidly fires the bolter gun!</span>")
+		face_atom(target)
+		for(var/i = 1 to 5)
+			new /obj/effect/temp_visual/dir_setting/firing_effect(loc, dir)
+			Shoot(target)
+			SLEEP_CHECK_DEATH(3)
```

Comment:
are you sure that you can't just use callbacks or something? sleep()s are generally bad news bears

Reply:
I am not sure on how to do it, and this variant works well enough.